New rag agent

thinkall opened the pull request:
<!-- Thank you for your contribution! Please review https://microsoft.github.io/autogen/docs/Contribute before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Refactor RAG agents. In this PR, only core functionalities are added. After we agree on the new architecture, enhancements could be added in other PRs.

Example Code:
```python
import autogen
from autogen.agentchat.contrib.rag import RagAgent, logger
import logging

logger.setLevel(logging.DEBUG)

config_list = autogen.config_list_from_json(
    "OAI_CONFIG_LIST",
    file_location=".",
    filter_dict={
        "model": ["gpt-3.5-turbo", "gpt-35-turbo", "gpt-35-turbo-0613", "gpt-4", "gpt4", "gpt-4-32k"],
    },
)

print("LLM models: ", [config_list[i]["model"] for i in range(len(config_list))])

llm_config = {
    "timeout": 60,
    "config_list": config_list,
}


def termination_msg(x):
    return isinstance(x, dict) and "TERMINATE" == str(x.get("content", ""))[-9:].upper()


userproxy = autogen.UserProxyAgent(
    name="userproxy",
    is_termination_msg=termination_msg,
    human_input_mode="ALWAYS",
    code_execution_config={"use_docker": False, "work_dir": ".tmp"},
    default_auto_reply="Reply `TERMINATE` if the task is done.",
    description="The boss who ask questions and give tasks.",
)


rag_config = {
    "docs_path": "./website/docs",
}

rag = RagAgent(
    name="rag",
    is_termination_msg=termination_msg,
    human_input_mode="NEVER",
    max_consecutive_auto_reply=5,
    llm_config=llm_config,
    rag_config=rag_config,
    code_execution_config=False,
    description="Assistant who has extra content retrieval power for solving difficult problems.",
)

userproxy.initiate_chat(recipient=rag, message="What is AutoGen?")
```

![new_rag](https://github.com/microsoft/autogen/assets/3197038/7f286f6a-e972-48c4-ab33-016bc1a4b9b7)

## Related issue number

Closes #1726 

## Tasks
- [x] Core functionalities
- [x] Notebook example
- [x] RAG capability 
- [ ] unstructured


## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

rickyloynd-microsoft commented:
Exciting!

It looks like this PR is the first step in replacing `RetrieveUserProxyAgent` and `RetrieveAssistantAgent` with a single `RagAgent` that can communicate with any `UserProxyAgent`. 

Does the roadmap #1657 include eventually making RAG a capability that can be added to any `ConversableAgent`?

thinkall commented:
> Exciting!
> 
> It looks like this PR is the first step in replacing `RetrieveUserProxyAgent` and `RetrieveAssistantAgent` with a single `RagAgent` that can communicate with any `UserProxyAgent`.
> 
> Does the roadmap #1657 include eventually making RAG a capability that can be added to any `ConversableAgent`?

Yes, that's one of the goal. I just added a task in the roadmap for tracking.

BeibinLi commented on autogen/agentchat/contrib/rag/vectordb.py:
```diff
@@ -0,0 +1,199 @@
+from .encoder import Encoder
+
+
+class VectorDB(ABC):
```
We decided to use `Protocol` rather than `ABC` in the future. See: https://jellis18.github.io/post/2022-01-11-abc-vs-protocol/

thinkall replied:
Done.

BeibinLi commented on autogen/agentchat/contrib/rag/chromadb.py:
```diff
@@ -0,0 +1,248 @@
+    raise ImportError("Please install chromadb: `pip install chromadb`")
+
+
+class ChromaVectorDB(VectorDB):
```
If we use `Protocol`, it does not need to inherit from VectorDB.

BeibinLi reviewed (commented):
This is a GREAT and big PR. I am only half-way through it. 

My feeling is that we need an additional README.md file under the "rag" folder to describe the design with something illustrative (such as UML diagram). Happy to help.

joshkyh commented on autogen/agentchat/contrib/rag/reranker.py:
```diff
@@ -0,0 +1,121 @@
+        self.vectorizer = TfidfVectorizer()
+        self.docs = None
+
+    def vecotrize(self, docs: List[str]) -> None:
```
```suggestion
    def vectorize(self, docs: List[str]) -> None:
```

thinkall replied:
Thanks for pointing out. Fixed.

joshkyh reviewed (commented):
Thanks for the contribution! Excited about the Abstract class for vector database, VectorDB.

joshkyh commented on autogen/agentchat/contrib/rag/reranker.py:
```diff
@@ -0,0 +1,121 @@
+            List[Union[int, Tuple[Union[int, str], float]]] | The reranked documents or the reranked documents with scores.
+        """
+        if docs and self.docs != docs:
+            self.vecotrize(docs)
```
```suggestion
            self.vectorize(docs)
```

joshkyh commented on test/agentchat/contrib/rag/test_reranker.py:
```diff
@@ -0,0 +1,66 @@
+        self.docs = ["This is document 1", "This is document 2", "This is document 3"]
+        self.query = Query("query", k=2)
+
+    def test_vecotrize(self):
```
```suggestion
    def test_vectorize(self):
```
vectorize

joshkyh commented on test/agentchat/contrib/rag/test_reranker.py:
```diff
@@ -0,0 +1,66 @@
+        self.query = Query("query", k=2)
+
+    def test_vecotrize(self):
+        self.reranker.vecotrize(self.docs)
```
```suggestion
        self.reranker.vectorize(self.docs)
```

joshkyh commented on test/agentchat/contrib/rag/test_reranker.py:
```diff
@@ -0,0 +1,66 @@
+        self.assertIsNotNone(self.reranker.corpus_tfidf)
+
+    def test_rerank_return_docs(self):
+        self.reranker.vecotrize(self.docs)
```
```suggestion
        self.reranker.vectorize(self.docs)
```

joshkyh commented on test/agentchat/contrib/rag/test_reranker.py:
```diff
@@ -0,0 +1,66 @@
+        self.assertIsInstance(reranked_docs[0], str)
+
+    def test_rerank_return_scores(self):
+        self.reranker.vecotrize(self.docs)
```
```suggestion
        self.reranker.vectorize(self.docs)
```

joshkyh commented on test/agentchat/contrib/rag/test_reranker.py:
```diff
@@ -0,0 +1,66 @@
+        self.assertIsInstance(reranked_docs[0][1], float)
+
+    def test_rerank_return_docs_and_scores(self):
+        self.reranker.vecotrize(self.docs)
```
```suggestion
        self.reranker.vectorize(self.docs)
```

skzhang1 commented:
Very large and impressive PR! I learned a lot after going through it. I think this interface is more elegant compared to the previous one, which consolidates RAG capabilities into a single RagAgent. Also happy to help if needed.

sonichi commented:
Do users always need to use a special RAG Agent, or can they add RAG capability to any agent, like teachability?

thinkall commented:
> Do users always need to use a special RAG Agent, or can they add RAG capability to any agent, like teachability?

RAG capability added.

rickyloynd-microsoft commented:
> > Do users always need to use a special RAG Agent, or can they add RAG capability to any agent, like teachability?
> 
> RAG capability added.

Awesome!

It looks like the code example in rag/README.md hasn't been updated yet to use the new capability.

thinkall commented:
> > > Do users always need to use a special RAG Agent, or can they add RAG capability to any agent, like teachability?
> > 
> > 
> > RAG capability added.
> 
> Awesome!
> 
> It looks like the code example in rag/README.md hasn't been updated yet to use the new capability.

Updated, also added a simple notebook example.

thinkall commented:
The docs/checks failing is not clear to me. @jackgerrits , do you know why it fails here? Thanks.
[agentchat_RAG_new.mdx.txt](https://github.com/microsoft/autogen/files/14418509/agentchat_RAG_new.mdx.txt)


![image](https://github.com/microsoft/autogen/assets/3197038/d60f1880-4270-45cc-ab3c-d62eccb355ec)

rickyloynd-microsoft commented:
> > > > Do users always need to use a special RAG Agent, or can they add RAG capability to any agent, like teachability?
> > > 
> > > 
> > > RAG capability added.
> > 
> > 
> > Awesome!
> > It looks like the code example in rag/README.md hasn't been updated yet to use the new capability.
> 
> Updated, also added a simple notebook example.

The notebook `agentchat_RAG_new.ipynb` gives two examples using a `RagAgent` and one example using an `AssistantAgent` to which `Ragability` was added. Is there any functional difference between these two agents?

ekzhu commented on autogen/agentchat/contrib/rag/rag_agent.py:
```diff
@@ -0,0 +1,810 @@
+        else:
+            return True, None if proxy_reply is None else proxy_reply["content"]
+
+    def run_code(self, code, **kwargs) -> Tuple[int, str, None]:
```
Is there a reason to run code in this Agent?

thinkall replied:
Not necessary, removed.

ekzhu commented on autogen/agentchat/contrib/rag/rag_agent.py:
```diff
@@ -0,0 +1,810 @@
+            is_termination_msg=lambda m: False,
+        )
+
+        self.ipython = get_ipython()
```
We currently have Jupyter code executor and adding the containerized version of it. That would be a more secure version of running code directly in the embedded ipython kernel. https://github.com/microsoft/autogen/pull/1794

thinkall commented:
> Spending a lot of time installing new packages. Would be nice to have a "requirements.txt" for pip install or equivalent. Also, could never get past the "ImportError("Please install chromadb: `pip install chromadb`")" even though I installed chromadb.

Thank you @radman-x for the feedback. The requirements is in defined in `setup.py`, extras_require `rag`.

Regarding `chromadb` import error. Could you try below lines one by one? If all good, the code inside autogen should be good as well.

```
    import chromadb

    if chromadb.__version__ < "0.4.15":
        raise ImportError("Please upgrade chromadb to version 0.4.15 or later.")
    from chromadb.api.models.Collection import Collection
```

jackgerrits commented:
The front_matter needs to be moved to the notebook metadata instead of comment in the notebook. See here: https://github.com/microsoft/autogen/blob/main/notebook/contributing.md

anu43 commented:
I was exploring different ways to implement vector databases without relying on RetrieveUserProxyAgent. My search began with the GitHub issue https://github.com/microsoft/autogen/issues/1657. This led me to explore the related issues and pull requests, specifically: https://github.com/microsoft/autogen/issues/725 > PR https://github.com/microsoft/autogen/pull/161 > PR https://github.com/microsoft/autogen/pull/619 > and https://github.com/microsoft/autogen/pull/967. Despite this, I couldn't find documentation on the RagAgent class in the AutoGen Docs. I'm curious whether `rag_config` relates or could be addressed to my goal in the code below.

```python
class CustomRetrieveAgent(RetrieveUserProxyAgent):
    def query_api(self, collection, query):
        # API call to retrieve chunks from Azure
        url = f"https://api.azure-api.com/collections/{collection}/query"
        headers = {"Authorization": "Bearer {{TOKEN}}"}
        data = {"question": query}

        response = requests.post(url, headers=headers, json=data, timeout=500)
        if response.status_code == 200:
            return response.json().get("context")[:5]  # Return top 5 chunks
        else:
            print(f"Error: {response.status_code}")
            return None

    def retrieve_docs(self, query: str, n_results: int = 20, **kwargs):
        api_results = self.query_api("default", query)
        
        results = []
        for i, chunk in enumerate(api_results):
            doc = {
                "id": chunk["id"],
                "metadata": chunk["metadata"],
                "content": chunk["content"],
            }
            relevance_score = 1.0 - (i / len(api_results))
            results.append((doc, relevance_score))

        self._results = [results]
```
